### PR TITLE
Implement auditor

### PR DESCRIPTION
**Pull request by jerry128371**
## Why
Implement auditor to accomplish BOLT protocol

## What
- Type1 filter (repeated GSN receipts)
- Type2 filter (wrong balance receipts, total amount > contract bond)
- Type2 filter (wrong balance receipts, total amount <= contract bond)
- Type4 filter (skipped GSN receipts)
- Type5 Filter (receipts without integrity)

### CODE REVIEW

File: src/auditor.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    let wrongSignature = [];
+    let type1 = [];
+    let type2 = [];
+    let type3 = [];
+    let type4 = [];
+    let type5 = [];
+    let bond = 1000000;// fetch bond by sidechain contract
+
+    // Arrange the receipts by GSN
+    let sequenceReceipts = [];
```
Comment (juinc) [inline src/auditor.js]: `let sequenceReceipts = stageReceipts.map(...)`
不需要先宣告空陣列

File: src/auditor.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      wrongSignature: wrongSignature,
+      type1: type1,
+      type2: type2,
+      type3: type3,
+      type4: type4,
+      type5: type5
+    };
+  }
+
+  _recover = (msgHash, signature) => {
```
Comment (juinc) [inline src/auditor.js]: use `verifier` instead

File: src/auditor.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+class Auditor {
+  constructor (auditorConfig, infinitechain) {
+    this._infinitechain = infinitechain;
+    this.serverAddress = auditorConfig.serverAddress;
+    this._storage = auditorConfig.storage;
+    this._nodeUrl = auditorConfig.nodeUrl;
+  }
+
+  filterWrongReceipts = async (stageReceipts, accounts) => { // previous stage accounts
```
Comment (juinc) [inline src/auditor.js]: 應該叫 `audit` ?

File: src/auditor.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      if (recoverServerReceiptAddress != this.serverAddress) {
+        wrongSignature.push({
+          wrongServerReceiptSig: receipt
+        });
+      }
+    });
+
+    // type1 double GSN
+    let existGSN = {};
+    sequenceReceipts.map(receipt => {
```
Comment (juinc) [inline src/auditor.js]: 1. use `forEach` instead of `map`
2. type1 ~ type5 為何不放在同一個 `forEach`

**Comment by juinc** [comment]
需補 test

File: src/auditor.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    let wrongSignature = [];
+    let type1 = [];
+    let type2 = [];
+    let type3 = [];
+    let type4 = [];
+    let type5 = [];
+    let existGSN = {};// type1 use
+    let gsnCounter = 1;// type4 use
+    let bond = 1000000;// fetch bond by sidechain contract
+    bond = bond.padStart(64, '0').slice(-64);
```
Comment (juinc) [inline src/auditor.js]: No need to convert to hex

File: src/auditor.js [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    bond = new BigNumber('0x' + bond);
+    // Arrange the receipts by GSN
+    let orderReceipts = stageReceipts.sort(function (a, b) {
+      return a.receiptData.GSN - b.receiptData.GSN;
```
Comment (juinc) [inline src/auditor.js]: `GSN` should be parsed to integer

File: src/auditor.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    let type3 = [];
+    let type4 = [];
+    let type5 = [];
+    let existGSN = {};// type1 use
+    let gsnCounter = 1;// type4 use
+    let bond = 1000000;// fetch bond by sidechain contract
+    bond = bond.padStart(64, '0').slice(-64);
+    bond = new BigNumber('0x' + bond);
+    // Arrange the receipts by GSN
+    let orderReceipts = stageReceipts.sort(function (a, b) {
```
Comment (juinc) [inline src/auditor.js]: order => ordered

File: src/auditor.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+      return a.receiptData.GSN - b.receiptData.GSN;
+    });
+
+    orderReceipts.forEach(receipt => {
+      // verify signature
+      let result = this._verifySignature(receipt, wrongSignature);
```
Comment (juinc) [inline src/auditor.js]: `wrongSignature` should be removed

File: src/auditor.js [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+      if (result == false) {
```
Comment (juinc) [inline src/auditor.js]: `if (!result)`

File: src/auditor.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        if (type2And3Result.value > bond) {
+          // wrong receipt value is larger than bond
+          type2.push(receipt);
+        } else {
+          // wrong receipt value is less than bond
+          type3.push(receipt);
+        }
+      }
+      // type4 continual GSN
+      let type4Result = this._type4Filter(receipt, gsnCounter, orderReceipts);
```
Comment (juinc) [inline src/auditor.js]: Calculate diff of two continuous receipts

File: src/auditor.js [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    let gsnCounter = 1;// type4 use
+    let bond = 1000000;// fetch bond by sidechain contract
+    bond = bond.padStart(64, '0').slice(-64);
+    bond = new BigNumber('0x' + bond);
+    // Arrange the receipts by GSN
+    let orderReceipts = stageReceipts.sort(function (a, b) {
+      return a.receiptData.GSN - b.receiptData.GSN;
+    });
+
+    orderReceipts.forEach(receipt => {
```
Comment (juinc) [inline src/auditor.js]: Use `reduce` since the final result is an object

File: src/auditor.js [hunk 12]
```diff
@@ -12,68 +12,62 @@ class Auditor {
...
+    bond = new BigNumber(bond);
     // Arrange the receipts by GSN
-    let orderReceipts = stageReceipts.sort(function (a, b) {
-      return a.receiptData.GSN - b.receiptData.GSN;
+    let orderedReceipts = stageReceipts.sort(function (a, b) {
+      return parseInt('0x' + a.receiptData.GSN) - parseInt('0x' + b.receiptData.GSN);
     });
 
-    orderReceipts.forEach(receipt => {
+    orderedReceipts.forEach(receipt => {
```
Comment (juinc) [inline src/auditor.js]: Use `reduce` to collect `objectedReceipts`

File: src/auditor.js [hunk 13]
```diff
@@ -12,68 +12,62 @@ class Auditor {
...
-    let gsnCounter = 1;// type4 use
     let bond = 1000000;// fetch bond by sidechain contract
-    bond = bond.padStart(64, '0').slice(-64);
-    bond = new BigNumber('0x' + bond);
+    bond = new BigNumber(bond);
     // Arrange the receipts by GSN
-    let orderReceipts = stageReceipts.sort(function (a, b) {
-      return a.receiptData.GSN - b.receiptData.GSN;
+    let orderedReceipts = stageReceipts.sort(function (a, b) {
+      return parseInt('0x' + a.receiptData.GSN) - parseInt('0x' + b.receiptData.GSN);
```
Comment (juinc) [inline src/auditor.js]: Use `parseInt(GSN, 16)` (cleaner)

File: test/test-auditor.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      setSignerKey('41b1a0649752af1b28b3dc29a1556eee781e4a4c3a1f7f53f90fa834de098c4d').
+      setStorage('memory').
+      build();
+
+    await infinitechain.initialize();
+  });
+
+  describe('#audit', () => {
+    it('returns correct result', async () => {
+      // await infinitechain.auditor.audit(stageHeight, receiptsWithRepeatedGSN, initialBalances, bond);
```
Comment (BenjaminLu) [inline test/test-auditor.js]: Should delete these comments.

**Pull request merged**